ddtrace/internal/_encoding.pyx

```diff
@@ -46,120 +47,327 @@ cdef inline int PyBytesLike_Check(object o):
+            if ret == 0:
+                ret = msgpack_pack_raw_body(pk, <char *> text, L)
+        return ret
+
+    raise TypeError("Unhandled text type: %r" % type(text))
+
 
-    usage::
+cdef inline int array_prefix_size(int l):
+    if l < 16:
+        return 1
+    elif l < (2<<16):
+        return 3
+    return 5
 
-        packer = Packer()
-        astream.write(packer.pack_trace(trace))
-        astream.write(packer.pack_traces(traces))
 
-    Packer's constructor has some keyword arguments:
+cdef class StringTable(object):
+    cdef dict _index
 
-    :param callable default:
-        Convert user type to builtin type that Packer supports.
-        See also simplejson's document.
-    """
+    def __init__(self):
+        self._index = {"": 0}
+        self.insert("")
+
+    cpdef insert(self, object string):
+        pass
+
+    cpdef index(self, object string):
+        cdef int index
+
+        if not string:
+            return 0
+        
+        try:
+            return self._index[string]
+        except KeyError:
+            index = len(self)
+            self._index[string] = index
+            self.insert(string)
+            return index
+
+    def __len__(self):
+        return len(self._index)
+
+    def __contains__(self, object string):
+        return string in self._index
+
+
+cdef class MsgpackStringTable(StringTable):
     cdef msgpack_packer pk
-    cdef object _default
-    cdef const char *encoding
-    cdef const char *unicode_errors
+    cdef int max_size
+    cdef int _sp
+    cdef object _lock
 
-    def __cinit__(self):
-        cdef int buf_size = 1024*1024
-        self.pk.buf = <char*> PyMem_Malloc(buf_size)
+    def __init__(self, max_size):
+        self.pk.buf = <char*> PyMem_Malloc(max_size)
         if self.pk.buf == NULL:
             raise MemoryError("Unable to allocate internal buffer.")
-        self.pk.buf_size = buf_size
-        self.pk.length = 0
+        self.max_size = self.pk.buf_size = max_size
+        self.pk.length = 6
+        self._sp = 0
+        self._lock = threading.Lock()
+        super(MsgpackStringTable, self).__init__()
+
+    def __dealloc__(self):
+        PyMem_Free(self.pk.buf)
+        self.pk.buf = NULL
 
-    def __init__(self, default=None):
-        if default is not None:
-            if not PyCallable_Check(default):
-                raise TypeError("default must be a callable.")
-        self._default = default
+    cpdef insert(self, object string):
+        cdef int ret
+
+        ret = pack_text(&self.pk, string)
+        if ret != 0:
+            raise RuntimeError("Failed to add string to msgpack string table")
+
+    cpdef savepoint(self):
+        self._sp = self.pk.length
+
+    cpdef rollback(self):
+        if self._sp > 0:
+            self.pk.length = self._sp
+
+    cpdef get_bytes(self):
+        cdef int l = len(self._index)
+        cdef int offset = 6 - array_prefix_size(l)
+        cdef int old_pos = self.pk.length
```

Comment:
Could you explain what's happening in `get_bytes`? A little lost 😕

Reply:
This is a bit of a weird thing to do, especially on an instance of `StringTable` and it depends on the v0.5 specs. The payload should look like an array of two objects: a string table and an array of traces. Since the string table comes before the array of traces, I thought of reusing the string table buffer so that we would only have to copy the array of traces and keep the string table in place. For this not only we need to update the size of the string table (which would require up to 5 bytes) but also add the "root" array prefix, which is of size 2, and therefore requires a single extra byte. Whence the 6.